servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl

```diff
@@ -427,106 +429,107 @@ void vertex_shader(vec3 vertex_input,
 	vec3 temp_precision; // Will be ignored.
 	vertex += double_add_vec3(model_origin, model_precision, scene_data.inv_view_matrix[3].xyz, view_precision, temp_precision);
 	vertex = mat3(scene_data.view_matrix) * vertex;
-#else
+#else // USE_DOUBLE_PRECISION
 	vertex = (modelview * vec4(vertex, 1.0)).xyz;
-#endif
+#endif // USE_DOUBLE_PRECISION
+
 #ifdef NORMAL_USED
 	normal = modelview_normal * normal;
-#endif
+#endif // NORMAL_USED
 
 #ifdef TANGENT_USED
-
 	binormal = modelview_normal * binormal;
 	tangent = modelview_normal * tangent;
-#endif
-#endif // !defined(SKIP_TRANSFORM_USED) && !defined(VERTEX_WORLD_COORDS_USED)
+#endif// TANGENT_USED
+
+#endif // !SKIP_TRANSFORM_USED && !VERTEX_WORLD_COORDS_USED
 
-//using world coordinates
+// Using world coordinates.
 #if !defined(SKIP_TRANSFORM_USED) && defined(VERTEX_WORLD_COORDS_USED)
 
 	vertex = (scene_data.view_matrix * vec4(vertex, 1.0)).xyz;
 #ifdef NORMAL_USED
 	normal = (scene_data.view_matrix * vec4(normal, 0.0)).xyz;
-#endif
+#endif // NORMAL_USED
 
 #ifdef TANGENT_USED
 	binormal = (scene_data.view_matrix * vec4(binormal, 0.0)).xyz;
 	tangent = (scene_data.view_matrix * vec4(tangent, 0.0)).xyz;
-#endif
-#endif
+#endif // TANGENT_USED
+
+#endif // !SKIP_TRANSFORM_USED && VERTEX_WORLD_COORDS_USED
 
 	vertex_interp = vertex;
 
+// Normalize TBN vectors before interpolation, per MikkTSpace.
+// See: http://www.mikktspace.com/
 #ifdef NORMAL_USED
-	normal_interp = normal;
-#endif
+	normal_interp = normalize(normal);
+#endif // NORMAL_USED
 
 #ifdef TANGENT_USED
-	tangent_interp = tangent;
-	binormal_interp = binormal;
-#endif
+	tangent_interp = normalize(tangent);
+	binormal_interp = normalize(binormal);
+#endif // TANGENT_USED
 
 #ifdef MODE_RENDER_DEPTH
 
 #ifdef MODE_DUAL_PARABOLOID
 
 	vertex_interp.z *= scene_data.dual_paraboloid_side;
 
-	dp_clip = vertex_interp.z; //this attempts to avoid noise caused by objects sent to the other parabolloid side due to bias
-
-	//for dual paraboloid shadow mapping, this is the fastest but least correct way, as it curves straight edges
+	dp_clip = vertex_interp.z; // This attempts to avoid noise caused by objects sent to the other parabolloid side due to bias.
```

Comment:
```suggestion
	dp_clip = vertex_interp.z; // This attempts to avoid noise caused by objects sent to the other paraboloid side due to bias.
```
Tiny correction 👀

Reply:
Appreciated. 👍